batch/batch/server/server.py

```diff
@@ -452,8 +448,8 @@ def __init__(self, id, batch_id, attributes, callback, userdata, user, always_ru
         if self.is_complete():
             return
         if self._state == 'Created':
-            self._cancelled = True
-            await db.jobs.update_record(self.id, cancelled=True)
+            if not self.always_run:
+                await self.set_state('Cancelled')
```

Comment:
Thinking through what this changes.

Suppose we cancel a pod. We hit this `await`, we may yield control to another coroutine. Because there is no cancelled flag, that other coroutine will not know we're about to cancel. It might start a pod. `set_state` does not destroy any pods. That means we can end up with a running pod but our state set to `"Cancelled"`.

Reply:
@cseed and I discussed this earlier during 1:1. We need to set the state first before destroying pods and we need two polling loops to catch edge cases where pods are created while the job is being cancelled. This is fine for now.